lib/mongo/server/connection_pool/queue.rb

```diff
@@ -313,39 +332,101 @@ def close_stale_sockets!
+
+          publish_cmap_event(
+            Monitoring::Event::ConnectionClosed.new(
+              reason,
+              @address,
+              connection.id,
+            )
+          )
+        end
+
+        def is_stale?(connection)
+          if connection.generation != @generation
+            close_connection!(connection, Monitoring::Event::ConnectionClosed::STALE)
+            true
+          end
+        end
+
+        def is_idle?(connection)
+          if connection && connection.last_checkin && max_idle_time
+            if Time.now - connection.last_checkin > max_idle_time
+              close_connection!(connection, Monitoring::Event::ConnectionClosed::IDLE)
+              true
+            end
+          end
+        end
+
         def dequeue_connection
+          semaphore = check_wait_queue
           deadline = Time.now + wait_timeout
-          loop do
-            return queue.shift unless queue.empty?
-            connection = create_connection
-            return connection if connection
-            wait_for_next!(deadline)
+          semaphore.wait(wait_timeout) if semaphore
+          raise Error::WaitQueueTimeout.new(@address, pool_size) if deadline <= Time.now
+
+          mutex.synchronize do
+            loop do
+              until queue.empty?
+                connection = queue.shift
+
+                unless is_stale?(connection) || is_idle?(connection)
+                  return connection
+                end
+              end
+
+              connection = create_connection
+              return connection if connection
+
+              wait = deadline - Time.now
+              resource.wait(mutex, wait)
+              raise Error::WaitQueueTimeout.new(@address, pool_size) if deadline <= Time.now
+            end
           end
         end
 
+        def check_wait_queue
+          semaphore = Semaphore.new
+
+          @wait_queue_mutex.synchronize do
+            @wait_queue << semaphore
+
+            # No need to wait for semaphore if nothing else is in the wait queue.
+            if @wait_queue.size == 1
+              return
+            end
+          end
+
+
+          semaphore
+        end
+
         def create_connection
           if pool_size < max_size
             @pool_size += 1
             @block.call(@generation)
           end
         end
 
-        def wait_for_next!(deadline)
-          wait = deadline - Time.now
-          if wait <= 0
-            raise Timeout::Error.new("Timed out attempting to dequeue connection after #{wait_timeout} sec.")
-          end
-          resource.wait(mutex, wait)
-        end
-
-        def check_count_invariants
+        # We only create new connections when we're below the minPoolSize on creation, when we're
+        # disconnecting and starting a new generation, and when we're checking out connections (per
+        # the CMAP spec), so `check_min` should be false in all other cases.
+        def check_count_invariants(check_min = true)
```

Comment:
Since connections are created on the foreground thread, in which circumstances is the min invariant violated by this PR as currently implemented?

Reply:
In all other situations than described by the comment to this method (i.e. wherever this method is called with `false` as the parameter). One example is when a pool with minPoolSize > 1 is cleared and then a connection is checked out; according to the spec, the connections beyond the first should be created in a non-blocking manner, so we can't create the rest of the connections needed to reach the min pool size until we implement the background thread, which you requested that we hold off for a separate review.

Reply:
The existing implementation (https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/server/connection_pool/queue.rb#L151) created replacement connection objects upon disconnecting existing connections, thus maintaining the invariants. What is preventing the new implementation from maintaining this behavior?

Reply:
Based on offline discussions, I think we've determined that the new connections should be created by the background thread. Is this good to resolve?